siliconcompiler/schema/schema_obj.py

```diff
@@ -194,204 +229,162 @@ def record_history(self):
+            if not isinstance(value, list):
+                value = [value]
+            base_type = sc_type[1:-1]
+            return [Schema._normalize_value(v, base_type) for v in value]
+
+        if sc_type.startswith('('):
+            # TODO: make parsing more robust to support tuples-of-tuples
+            if isinstance(value, str):
+                value = value[1:-1].split(',')
+            elif not (isinstance(value, tuple) or isinstance(value, list)):
+                raise TypeError
+
+            base_types = sc_type[1:-1].split(',')
+            if len(value) != len(base_types):
+                raise TypeError
+            return tuple(Schema._normalize_value(v, base_type) for v, base_type in zip(value, base_types))
+
+        if sc_type == 'bool':
+            if value == 'true': return True
+            if value == 'false': return False
+            if isinstance(value, bool): return value
+            raise TypeError
+
+        if sc_type == 'int':
+            return int(value)
+
+        if sc_type == 'float':
+            return float(value)
+
+        if sc_type in ('str', 'file', 'dir'):
+            if isinstance(value, str): return value
+            else: raise TypeError
+
+        raise ValueError(f'Invalid type specifier: {sc_type}')
+
+    @staticmethod
+    def _normalize_field(value, sc_type, field):
+        if Schema._is_list(field, sc_type):
+            if not isinstance(value, list):
+                value = [value]
+            if not all(isinstance(v, str) for v in value):
+                raise TypeError(f'Invalid value {value} for {field}: expected str')
+            return value
+
+        if field == 'scope':
+            # Restricted allowed values
+            if not (isinstance(value, str) and value in ('global', 'job', 'scratch')):
+                raise TypeError(f'Invalid value {value} for field {field}: '
+                    'expected one of "global", "job", or "scratch"')
+            return value
+
+        if field in (
+            'type', 'switch', 'shorthelp', 'help', 'unit', 'hashalgo', 'notes',
+            'signature'
+        ):
+            if not isinstance(value, str):
+                raise TypeError(f'Invalid value {value} for field {field}: expected str')
+            return value
+
+        if field in ('require', 'lock', 'copy', 'set'):
+            if value == 'true': return True
+            if value == 'false': return False
+            if isinstance(value, bool): return value
+            else: raise TypeError(f'Invalid value {value} for field {field}: expected bool')
+
+        raise ValueError(f'Invalid field: {field}')
+
+    @staticmethod
+    def _is_empty(cfg):
+        return not cfg['set']
+
+    @staticmethod
+    def _is_leaf(cfg):
+        return 'shorthelp' in cfg and isinstance(cfg['shorthelp'], str)
+
+    @staticmethod
+    def _is_list(field, type):
+        is_list = type.startswith('[')
+
+        if field in ('filehash', 'date', 'author', 'example'):
```

Comment:
Having these fields hard coded deep in the core/search function was always a bit annoying when we made a change in the schema to add a leaf field, but given how seldom we do that, I think this is fine.